src/core/ext/filters/client_channel/subchannel.cc

```diff
@@ -202,6 +202,7 @@ class ConnectedSubchannelStateWatcher
   // Must be instantiated while holding c->mu.
   explicit ConnectedSubchannelStateWatcher(grpc_subchannel* c)
       : subchannel_(c) {
+    gpr_mu_init(&mu_);
```

Comment:
I'm not keen on adding a new mutex here, since this class already holds the subchannel's mutex in various places (look for `c->mu`), and I don't want to introduce the possibility of a deadlock between the two.

Instead, how about the following:

- Add a `gpr_atm shutdown_` member.
- Set `shutdown_` to true in `Orphan()`.
- In `OnHealthChanged()`, when we check that the reported status is `GRPC_CHANNEL_SHUTDOWN`, also check the value of `shutdown_`.  If either one is true, then we do the same thing.

Reply:
I don't understand why this can solve the race.

If `Orphan()` is called after we check `shutdown_` and before we access `health_check_client_`, won't a segv still occur?

Reply:
Yeah, I think you're right.

So I guess we need a mutex of some sort here.  I would still prefer to avoid introducing a new one if possible, because I fear deadlocks between the two.  See if you can find a reasonable way to make use of the existing `c->mu` for this.  I suspect that this may be challenging, since `Orphan()` may be called in a context where we're already holding `c->mu`.  But if that's the case, then maybe we don't need to acquire it inside of `Orphan()` anyway.